std/algorithm/iteration.d

```diff
@@ -4280,3 +4284,237 @@ unittest
+if (isInputRange!R &&
+    !isInfinite!R &&
+    is(ElementType!R : double))
+in
+{
+    assert(!r.empty);
+}
+body
+{
+    static if (hasLength!R)
+    {
+        return sum(r) / r.length;
+    }
+    else
+    {
+        import std.typecons : tuple;
+
+        auto pair = reduce!((a, b) => tuple(a[0] + 1, a[1] + b))
+            (tuple(size_t(0), 0.0), r);
+        return pair[1] / pair[0];
+    }
+}
+
+///
+@safe unittest
+{
+    import std.math : approxEqual;
+
+    assert(mean([1, 2, 3]) == 2);
+    assert(approxEqual(mean([1.5, 2.5, 12.5]), 5.5));
+}
+
+unittest
+{
+    import std.internal.test.dummyrange : ReferenceInputRange;
+    import std.math : approxEqual;
+
+    auto r1 = new ReferenceInputRange!int([1, 2, 3]);
+    assert(r1.mean == 2);
+
+    auto r2 = new ReferenceInputRange!double([1.5, 2.5, 12.5]);
+    assert(approxEqual(r2.mean, 5.5));
+}
+
+/**
+Function to compute the quantile of a $(D SortedRange) given $(D percentile).
+Is $(D O(r.length * percentile)).
+
+Params:
+    r = a forward range with length
+    percentile = where to calculate the quantile
+
+Returns:
+    the quantile of the range given $(D percentile)
+*/
+double quantile(alias less = "a < b", R)(SortedRange!(R, less) r, double percentile) if (
+    isForwardRange!R &&
+    hasLength!R &&
+    !isRandomAccessRange!R &&
+    !isInfinite!R &&
+    is(ElementType!R : double))
+in
+{
+    assert(!r.empty);
+}
+body
+{
+    import std.math : modf;
+
+    if (r.length == 0)
+    {
+        return r.front;
+    }
+
+    real j;
+    real g;
+    double x = r.length * percentile;
+    g = modf(x, j);
+
+    if (g == 0)
+    {
+        ElementType!R prev_item;
+
+        foreach (index, item; r.enumerate)
+        {
+            if (index == cast(ulong) j - 1)
+            {
+                prev_item = item;
+            }
+            else if (index == cast(ulong) j)
+            {
+                return (item + prev_item) / 2.0;
+            }
+        }
+    }
+    else
+    {
+        foreach (index, item; r.enumerate)
+        {
+            if (index == cast(ulong) j)
+            {
+                return item;
+            }
+        }
+    }
+
+    assert(0);
+}
+
+/**
+Function to compute the quantile of a $(D SortedRange) given $(D percentile).
+Is $(D O(1)).
+
+Params:
+    r = a random access range
+    percentile = where to calculate the quantile
+
+Returns:
+    the quantile of the range given $(D percentile)
+*/
+double quantile(alias less = "a < b", R)(SortedRange!(R, less) r, double percentile) if (
+    isRandomAccessRange!R &&
+    !isInfinite!R &&
+    is(ElementType!R : double))
+in
+{
+    assert(!r.empty);
+}
+body
+{
+    import std.math : modf;
+
+    if (r.length == 0)
+    {
+        return r.front;
+    }
+
+    real j;
+    real g;
+    double x = r.length * percentile;
+    g = modf(x, j);
+
+    if (g == 0)
+    {
+        return (r[cast(ulong) j - 1] + r[cast(ulong) j]) / 2.0;
+    }
+    else
+    {
+        return r[cast(ulong) j];
```

Comment:
Are you sure you should be using casts here? What about using `std.math.round` or `std.math.trunc`?